### ISSUE TITLE
Y24-044 Remove 'asset' terminology

### ISSUE THREAD

**Issue by KatyTaylor** [sanger/sequencescape#4072]
**Describe the Housekeeping**
The word 'asset' is used throughout Sequencescape and is confusing / non-intuitive as it refers to a historical concept in the product.

Task: Research an approach to remove the word asset from the product (apart from a few exceptions, noted below), and create stories to do so.

A couple of example approaches / stories I can think of:
- By database vs code: Re-name all the foreign keys called `asset_id` on database tables at once (then do code references as a separate story).
- By table / object: Take one database table that has a column called `asset_id`, and rename that column and any reference to it in the code (then repeat with other tables, one by one).

**History:**
There used to be a table / model called Asset. Several years ago is was split into Labware (a piece of plastic used in the lab) and Receptacle (the hole(s) in that plastic that holds samples). In the database, there is a table called labware and a table called receptalces, but no table called assets. There is still a class called Asset, but it has minimal functionality. Wherever you find the word `asset` in the code, or as part of the name of a database table or column, it normally means `receptacle`; however, occasionally it means `labware`.

**Exceptions:**
This is a exercise in re-naming terminology within Sequencescape. To prevent scope creep, I think we should leave the following alone for now:
- Anything referring to Asset Audits - this is the name of another app, so references to this are OK
- The actual class called `Asset` - this story is about re-naming, not modifying functionality, so leave the name of this class and any references to it alone.

**Additional context**

This came out of a retrospective on 27/3/2024.

Data model:

![Assets, Labware   Receptacles](https://github.com/sanger/sequencescape/assets/10407809/c13a478a-00b0-4146-9ec8-bb82082d1faa)

**Comment by KatyTaylor** [comment]
**A possible approach:**

Considerations:

- Focus first on the tables that have asset_id as a foreign key. I think these cause the most confusion because you have to work out which table they link to (Labware or Receptacle).
- Do one table / model first (e.g. `barcodes`), then reassess the approach.
- Run database migration past the DBAs before running it in prod.
- Out of scope - changing app names (e.g. Asset Audits) or any repos outside Sequencescape (e.g. Limber or Sequencescpae Client API).

Steps:
1. Add an alias ‘labware_id’ / `receptacle_id` to the . Will use ‘asset_id’ column behind the scenes.
2. Update code at your leisure to use the new attribute name.
3. Disable the use of the old attribute name in code (e.g. override method, raise exception). This should catch any remaining references and be easier to roll back (ideally feature flag) than a db migration.
4. Rename column and remove alias.

Uncertainties:
- How to make sure we get all the code references?
- Is an alias equivalent to a database field for all types of code reference?
- alias vs alias_method vs alias_attribute

**Comment by KatyTaylor** [comment]
Testing out using `alias_attribute` as a stepping stone. Using the `Barcode` model.

Added this to the `Barcode` model:
```
# needed to add 2 aliases, so that both labware and labware_id could be successfully used.
alias_attribute :labware_id, :asset_id
alias_attribute :labware, :asset
```

Tested retrieving the linked labware using the above methods (the 'getter'):
```
plate = Purpose.find_by(name: 'Stock Plate').create!
barcode = plate.barcodes.first

# using the existing / old way
barcode.asset_id
barcode.asset

# using the aliases / new way
barcode.labware_id 
barcode.labware
```

Tested linking the labware upon creation, using the above methods (the 'setter'):
```
plate = Purpose.find_by(name: 'Stock Plate').create!
barcode_format = 7 # use fluidx foreign barcode for testing


# using create!
Barcode.create!(labware: plate, barcode:'MA21052401', format: barcode_format)
Barcode.create!(labware_id: plate.id, barcode: 'MA21052402', format: barcode_format)

# using create
Barcode.create(labware: plate, barcode: 'MA21052403', format: barcode_format)
Barcode.create(labware_id: plate.id, barcode: 'MA21052404', format: barcode_format)

# using new then save!
Barcode.new(labware: plate, barcode: 'MA21052405', format: barcode_format).save!
Barcode.new(labware_id: plate.id, barcode: 'MA21052406', format: barcode_format).save!

plate.barcodes.count # should be 7
```

All the above test code worked as expected.

**Comment by KatyTaylor** [comment]
Testing overriding `asset` / `asset_id` methods as a stepping stone to renaming the fields (in addition to aliases created in previous comment):

```
def asset_id
    raise "Use 'labware_id' instead. 'asset_id' is deprecated."
  end

  def asset
    raise "Use 'labware' instead. 'asset' is deprecated."
  end
```

This does not work, because when you use the aliases, Rails uses the original field names behind the scenes, and therefore accesses these methods. e.g.
```
plate = Plate.create!

Barcode.create!(labware: plate, barcode:'MA21052408', format: 7)
=> RuntimeError: Use 'labware' instead. 'asset' is deprecated.
```

**Comment by KatyTaylor** [comment]
The output of this story is the following 2 stories, which will tackle the first of many models!

- https://github.com/sanger/sequencescape/issues/4121
- https://github.com/sanger/sequencescape/issues/4123

The plan is to do these first, and reassess the approach. Then create stories for the next model.

**Comment by KatyTaylor** [comment]
Hi @StephenHulme 

Thanks for looking!

I made a hopefully comprehensive list of the affected tables in an earlier comment in this issue - https://github.com/sanger/sequencescape/issues/4072#issuecomment-2098445620

And a recommendation in this comment - https://github.com/sanger/sequencescape/issues/4072#issuecomment-2098455567 - to focus first on the tables that have a foreign key called `asset_id`, which will be a subset of the list above. 

I found looking through the schema.rb file to be v helpful in discovering affected tables.

Katy
